### ISSUE TITLE
Sketch compilation workflow templates incompatible with version 4.x of `actions/upload-artifact` dependency

### ISSUE THREAD

**Issue by per1234** [per1234/.github#109]
### Describe the problem

The [workflow templates](https://github.com/per1234/.github/tree/main/workflow-templates) hosted in this repository include a collection of workflows for "smoke testing" of Arduino projects by compiling sketches.

It is common for Arduino projects to target multiple boards. In this case, the workflow should be configured to compile for each of the target boards. The workflows are designed to accommodate this in an easily maintainable and efficient manner by using a [matrix](https://docs.github.com/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix) to generate a parallel job for the compilations for each board.

The workflows are intended to be used in combination with [the **arduino/report-size-deltas** action](https://github.com/arduino/report-size-deltas). For this reason, the sketch compilation job is configured to generate a [sketches report file](https://github.com/arduino/compile-sketches#sketches-report-path) and then upload it to a [workflow artifact](https://docs.github.com/actions/using-workflows/storing-workflow-data-as-artifacts) in order to transfer the data to the job or workflow where the **arduino/report-size-deltas** action will consume it and generate a "size deltas report" comment from the data.

The sketch compilation job is configured so that each matrix job uploads its sketches report file to a single artifact:

- [`workflow-templates/compile-sketches.yml#L36-L37`](https://github.com/per1234/.github/blob/2a587052e80aa1b0321645c72ddc4a6b813bf4ed/workflow-templates/compile-sketches.yml#L36-L37">workflow-templates/compile-sketches.yml#L36-L37)
- [`workflow-templates/compile-sketches.yml#L71-L76`](https://github.com/per1234/.github/blob/2a587052e80aa1b0321645c72ddc4a6b813bf4ed/workflow-templates/compile-sketches.yml#L71-L76)

```yaml
    env:
      SKETCHES_REPORTS_PATH: sketches-reports

[...]

      - name: Save sketches report as workflow artifact
        uses: actions/upload-artifact@v3
        with:
          if-no-files-found: error
          path: ${{ env.SKETCHES_REPORTS_PATH }}
          name: ${{ env.SKETCHES_REPORTS_PATH }}
```

Starting from version 4.0.0 of the **actions/upload-artifact** action, uploading multiple times to a single artifact is not possible:

https://github.com/actions/upload-artifact#breaking-changes

> 2. Uploading to the same named Artifact multiple times.
>
> Due to how Artifacts are created in this new version, it is no longer possible to upload to the same named Artifact multiple times. You must either split the uploads into multiple Artifacts with different names, or only upload once. Otherwise you will encounter an error.

This causes the **actions/upload-artifact** step of the workflow templates to fail if the action dependency is updated (https://github.com/per1234/.github/pull/107):

```text
Run actions/upload-artifact@v4
With the provided path, there will be 1 file uploaded
Artifact name is valid!
Root directory input is valid!
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run
```

The best way to adapt to this change would be adjusting the workflow configuration so that each sketch compilation matrix job uploads its sketches report file to a different workflow artifact (resulting in multiple  artifacts):

```yaml
  build:
    name: ${{ matrix.board.fqbn }}
    runs-on: ubuntu-latest
    permissions: {}

    env:
      SKETCHES_REPORTS_PATH: sketches-reports

    strategy:
      fail-fast: false

      matrix:
        board:
          # TODO: Add list of boards to compile for here:
          - fqbn: arduino:avr:uno
            sketches-report-artifact-suffix: arduino-avr-uno
            # TODO: Define the board's platform dependencies here.
            # See: https://github.com/arduino/compile-sketches#platforms
            platforms: |
              - name: arduino:avr

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Compile examples
        uses: arduino/compile-sketches@v1
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
          fqbn: ${{ matrix.board.fqbn }}
          platforms: ${{ matrix.board.platforms }}
          libraries: |
            # Disable the default behavior of installing the repository as a library.
            -
            # Additional library dependencies can be listed here.
            # See: https://github.com/arduino/compile-sketches#libraries
          sketch-paths: |
            - ${{ github.workspace }}
          enable-deltas-report: true
          sketches-report-path: ${{ env.SKETCHES_REPORTS_PATH }}

      - name: Save sketches report as workflow artifact
        uses: actions/upload-artifact@v4
        with:
          if-no-files-found: error
          path: ${{ env.SKETCHES_REPORTS_PATH }}
          name: sketches-report-${{ matrix.board.sketches-report-artifact-suffix }}
```

#### Public repository variants

In order to provide size deltas reports for pull requests from forks, the variants of the workflow templates for use in public repositories are intended to be used in combination with the ["**Report Size Deltas**" workflow template](https://github.com/per1234/.github/blob/main/workflow-templates/report-size-deltas.md). This workflow is triggered by a [`schedule` event](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onschedule). In this usage, the **arduino/report-size-deltas** action downloads the artifact containing the sketches reports produced by the sketch compilation workflow:

https://github.com/arduino/report-size-deltas#run-from-a-scheduled-workflow

The **arduino/report-size-deltas** action currently only supports downloading a single sketches report artifact and so is incompatible with the solution described above of creating a separate artifact for each sketch compilation matrix job. This is tracked at https://github.com/arduino/report-size-deltas/issues/55.

For this reason, it will be impossible to update the **actions/upload-artifact** dependency in the variants of the sketch compilation workflow templates for use in public repositories until the changes have been made in the **arduino/report-size-deltas** action codebase to allow it to consume sketches report data from multiple artifacts.

#### Private repository variants

Since the ["**Report Size Deltas**" workflow template](https://github.com/per1234/.github/blob/main/workflow-templates/report-size-deltas.md) is not well suited for use in private repositories (where GitHub Actions usage is [charged per-workflow run minute](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#per-minute-rates)), there are separate variants of the sketch compilation workflow templates designed for use in private repositories ([example](https://github.com/per1234/.github/blob/main/workflow-templates/compile-sketches-private.md)). In these workflows, where the **arduino/report-size-deltas** action runs in a workflow triggered by a [`pull_request` event](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request), the workflow uses the **actions/download-artifacts** action to download the sketches report artifact and the **arduino/report-size-deltas** action consumes the reports from the folder produced by the **actions/download-artifacts** step:

https://github.com/arduino/report-size-deltas#run-from-the-same-workflow-as-the-arduinocompile-sketches-action

In these workflow templates, the behavior of the **arduino/report-size-deltas** action is not directly relevant to the bump of the **actions/upload-artifact** dependency since **arduino/report-size-deltas** is not working directly with the artifact.

Unfortunately the **actions/upload-artifact** bump does still cause a problem for the size deltas report job even in these workflows. The reason is that the existing **actions/download-artifacts** step only downloads a single artifact, whereas the job must now download multiple artifacts. Intuitively, we would expect the solution to be using an artifact name pattern in the **actions/download-artifacts** step to download all matching artifacts:

```yaml
- name: Download sketches reports artifact
  id: download-artifact
  continue-on-error: true # If compilation failed for all boards then there are no artifacts
  uses: actions/download-artifact@v4
  with:
    name: sketches-report-*
    path: ${{ env.SKETCHES_REPORTS_PATH }}
```

Unfortunately **actions/download-artifacts** does not support the use of artifact name patterns: https://github.com/actions/download-artifact/issues/6. This means a separate step must be used for each artifact to be downloaded by the action:

```yaml
- name: Download sketches reports artifact
  uses: actions/download-artifact@v4
  with:
    name: sketches-report-arduino-avr-uno
    path: ${{ env.SKETCHES_REPORTS_PATH }}

- name: Download sketches reports artifact
  uses: actions/download-artifact@v4
  with:
    name: sketches-report-arduino-samd-mkrzero
    path: ${{ env.SKETCHES_REPORTS_PATH }}
```

This code duplication makes the workflow more difficult to install and maintain. Previously the only project-specific configuration required was the `jobs.build.strategy.matrix.board` array. With the strategy of using a dedicated **actions/download-artifact** step for each of the sketches report artifacts, it will be necessary for the template installer to add each of those steps, and for the maintainer to make sure these steps are kept synchronized with the contents of `jobs.build.strategy.matrix.board` as board targets are added or removed over time. For this reason, it would be worth investigating whether a better mechanism for downloading the multiple artifacts is available. If not, consideration might be given to changing the **arduino/report-size-deltas** action so that it can handle the artifact download in this use case, just as it already does in [the `schedule` event-triggered use case](https://github.com/arduino/report-size-deltas#run-from-a-scheduled-workflow).

### To reproduce

Update the version of the **actions/upload-artifact** action to `v4` in any workflow based on the sketch compilation workflow templates.

🐛 The **actions/upload-artifact** step in the workflow fails:

```text
Run actions/upload-artifact@v4
With the provided path, there will be 1 file uploaded
Artifact name is valid!
Root directory input is valid!
Error: Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run
```

### Expected behavior

Workflow templates are designed to be compatible with the latest release version of their action dependencies.

### Workflow templates version

2a587052e80aa1b0321645c72ddc4a6b813bf4ed

### Additional context

The inability to bump the `actions/upload-artifact` dependency also blocks the update of the `actions/download-artifact` dependency (https://github.com/per1234/.github/pull/108) in the private repository variants of the sketch compilation workflow templates:

https://github.com/actions/download-artifact#breaking-changes

> 2. Downloading artifacts that were created from `action/upload-artifact@v3` and below are not supported.

If **actions/download-artifact** is bumped without also bumping **actions/upload-artifact**, the download step will fail:

```text
Run actions/download-artifact@v4
Downloading single artifact
Error: Unable to download artifact(s): Artifact not found for name: sketches-reports
```

<a name="workaround"></a>

#### Workaround

Continue to use version 3.x of the **actions/upload-artifact** and **actions/download-artifact** (in the case of the private repo variants) dependencies in the sketch compilation workflow, as shown in [the current workflow templates](https://github.com/per1234/.github/tree/main/workflow-templates).

If you receive a pull request from Dependabot to bump the actions, it must not be accepted.

### Issue checklist

- [X] I searched for previous reports in [the issue tracker](https://github.com/per1234/.github/issues?q=)
- [X] I verified the problem still occurs when using the workflow template from the latest version
- [X] My report contains all necessary details

**Comment by per1234** [comment]
> The arduino/report-size-deltas action currently only supports downloading a single sketches report artifact

This has been resolved by https://github.com/arduino/report-size-deltas/pull/78, released in `arduino/report-size-deltas@v1.1.0`

> actions/download-artifacts does not support the use of artifact name patterns

This has been resolved by https://github.com/actions/download-artifact/pull/259, released in `actions/download-artifact@4.1.0`

**Comment by per1234** [comment]
Resolved by https://github.com/per1234/.github/pull/107
